File: spacevil_21052019/java_src/ToolTip.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;

import java.awt.Font;
import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import javax.swing.Timer;

public final class ToolTip extends Prototype {

    private static int count = 0;

    private TextLine _text_object;

    public TextLine getTextLine() {
        return _text_object;
    }

    private Timer _stop = null;
    private int _timeout = 500;

    public void setTimeOut(int milliseconds) {
        _timeout = milliseconds;
    }

    public int getTimeOut() {
        return _timeout;
    }

    // private static ToolTip _instance = null;

    protected ToolTip() {
        setVisible(false);
        _text_object = new TextLine();
        setItemName("ToolTip_" + count);
        count++;
        isFocusable = false;

        setStyle(DefaultsService.getDefaultStyle(ToolTip.class));
    }

    // public static ToolTip getInstance() {
    // if (_instance == null)
    // _instance = new ToolTip();
    // return _instance;
    // }

    @Override
    public void initElements() {
        // addItem(_text_object);
    }

    private Lock locker = new ReentrantLock();

    public void initTimer(boolean value) {
        locker.lock();
        try {
            if (value) {
                if (_stop != null)
                    return;
                ActionListener taskPerformer = new ActionListener() {
                    public void actionPerformed(ActionEvent evt) {
                        visibleSelf();
                    }
                };
                _stop = new Timer(_timeout, taskPerformer);
                _stop.start();
            } else {
                setVisible(value);
                if (_stop == null)
                    return;
                _stop.stop();
                _stop = null;
            }
        } finally {
            locker.unlock();
        }
    }

    private void visibleSelf() {
        locker.lock();
        try {
            if (_stop == null)
                return;
            setVisible(true);
            _stop.stop();
            _stop = null;
        } finally {
            locker.unlock();
        }
    }

    public int getTextWidth() {
        return _text_object.getWidth();
    }

    // text init
    public void setTextAlignment(List<ItemAlignment> alignment) {
        _text_object.setTextAlignment(alignment);
    }

    public void setTextMargin(Indents margin) {
        _text_object.setMargin(margin);
    }

    public void setFont(Font font) {
        _text_object.setFont(font);
    }

    public void setFontSize(int size) {
        _text_object.setFontSize(size);
    }

    public void setFontStyle(int style) {
        _text_object.setFontStyle(style);
    }

    public void setFontFamily(String font_family) {
        _text_object.setFontFamily(font_family);
    }

    public Font getFont() {
        return _text_object.getFont();
    }

    public void setText(String text) {
        _text_object.setItemText(text);
    }

    public String getText() {
        return _text_object.getItemText();
    }

    public void setForeground(Color color) {
        _text_object.setForeground(color);
    }

    public void setForeground(int r, int g, int b) {
        _text_object.setForeground(r, g, b);
    }

    public void setForeground(int r, int g, int b, int a) {
        _text_object.setForeground(r, g, b, a);
    }

    public void setForeground(float r, float g, float b) {
        _text_object.setForeground(r, g, b);
    }

    public void setForeground(float r, float g, float b, float a) {
        _text_object.setForeground(r, g, b, a);
    }

    public Color getForeground() {
        return _text_object.getForeground();
    }

    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);
        setForeground(style.foreground);
        setFont(style.font);
        setTextAlignment(style.textAlignment);
    }
}

File: spacevil_21052019/net_src/Items/ToolTip.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Timers;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    internal class ToolTip : Prototype, IToolTip
    {
        private TextLine _text_object;
        internal TextLine GetTextLine()
        {
            return _text_object;
        }
        internal Timer _stop;
        private int _timeout = 300;
        public void SetTimeOut(int milliseconds)
        {
            _timeout = milliseconds;
        }
        public int GetTimeOut()
        {
            return _timeout;
        }

        // private static ToolTip _instance = null;

        internal ToolTip()
        {
            SetVisible(false);
            _text_object = new TextLine();
            SetItemName("ToolTip");
            IsFocusable = false;

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ToolTip)));
        }

        // public static ToolTip GetInstance()
        // {
        //     if (_instance == null)
        //         _instance = new ToolTip();
        //     return _instance;
        // }

        public void SetTextAlignment(ItemAlignment alignment)
        {
            _text_object.SetTextAlignment(alignment);
        }
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            _text_object.SetTextAlignment(alignment);
        }
        public void SetFont(Font font)
        {
            _text_object.SetFont(font);
        }
        public void SetFontSize(int size)
        {
            _text_object.SetFontSize(size);
        }
        public void SetFontStyle(FontStyle style)
        {
            _text_object.SetFontStyle(style);
        }
        public void SetFontFamily(FontFamily font_family)
        {
            _text_object.SetFontFamily(font_family);
        }
        public Font GetFont()
        {
            return _text_object.GetFont();
        }
        public void SetText(String text)
        {
            _text_object.SetItemText(text);
        }
        public String GetText()
        {
            return _text_object.GetItemText();
        }
        public int GetTextWidth()
        {
            return _text_object.GetWidth();
        }
        public void SetForeground(Color color)
        {
            _text_object.SetForeground(color);
        }
        public Color GetForeground()
        {
            return _text_object.GetForeground();
        }

        internal void InitTimer(bool value)
        {
            if (value)
            {
                if (_stop != null)
                    return;

                _stop = new System.Timers.Timer(_timeout);
                _stop.Elapsed += (sender, e) => VisibleSelf();
                _stop.Start();
            }
            else
            {
                SetVisible(false);

                if (_stop == null)
                    return;

                _stop.Stop();
                _stop.Dispose();
                _stop = null;
            }
        }

        private void VisibleSelf()
        {
            SetVisible(true);

            _stop.Stop();
            _stop.Dispose();
            _stop = null;
        }

        //style
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);
            SetForeground(style.Foreground);
            SetFont(style.Font);
            SetTextAlignment(style.TextAlignment);
        }
    }
}